src/ASM/ASMs3D.C

```diff
@@ -1978,52 +1978,21 @@ bool ASMs3D::integrate (Integrand& integrand,
   bool use2ndDer = integrand.getIntegrandType() & Integrand::SECOND_DERIVATIVES;
   bool useElmVtx = integrand.getIntegrandType() & Integrand::ELEMENT_CORNERS;
 
-  // Get Gaussian quadrature points and weights
-  std::array<int,3> ng;
-  std::array<const double*,3> xg, wg;
-  for (int d = 0; d < 3; d++)
-  {
-    ng[d] = this->getNoGaussPt(svol->order(d));
-    xg[d] = GaussQuadrature::getCoord(ng[d]);
-    wg[d] = GaussQuadrature::getWeight(ng[d]);
-    if (!xg[d] || !wg[d]) return false;
-  }
+  if (myCache.empty())
+    myCache.emplace_back(std::make_unique<BasisFunctionCache>(*this, integrand, 1));
```

Comment:
What is emblace_back doing here? Same as push_back? If not, I lost the logic...
Also, the make_unique calls for some explanation. All these newer C++ construct I'm not updated with.

Reply:
emplace_back is used to construct in-place. with a push_back you first construct an object, and then copy that into the vector element. with emplace_back() this copy is avoided. make_unique creates a unique pointer in-place, avoiding default construction followed by assignment from bare pointer. so the sum effect is that an unique pointer is placed at the back of the vector without any copying.

Reply:
Yes. Strictly speaking this is not a shared_ptr, this is a unique_ptr. They are not copyable, only a single reference can exist, hence the requirement to create in-place.